src/Orleans.Core/Networking/ConnectionManager.cs

```diff
@@ -53,158 +55,162 @@ public int ConnectionCount
 
         public ValueTask<Connection> GetConnection(SiloAddress endpoint)
         {
-            if (this.connections.TryGetValue(endpoint, out var entry) && entry.Connections.Length >= this.connectionOptions.ConnectionsPerEndpoint)
+            if (!this.connections.TryGetValue(endpoint, out var entry)
+                || !entry.TryGetNextConnection(out var connection))
             {
-                var result = entry.Connections;
-                nextConnection = (nextConnection + 1) % result.Length;
-                var connection = result[nextConnection];
-                if (connection.IsValid) return new ValueTask<Connection>(connection);
+                return new ValueTask<Connection>(this.GetConnectionAsync(endpoint));
             }
 
-            return GetConnectionAsync(endpoint);
-        }
-
-        private async ValueTask<Connection> GetConnectionAsync(SiloAddress endpoint)
-        {
-            ImmutableArray<Connection> result;
-            ConnectionEntry entry = default;
-            var acquiredConnectionLock = false;
-            try
+            if (entry.Connections.Length < this.connectionOptions.ConnectionsPerEndpoint)
             {
-                // Lock the entry to ensure it will not be removed while the connectio attempt is occuring.
-                while (!acquiredConnectionLock)
+                lock (this.lockObj)
                 {
-                    entry = GetOrCreateEntry(endpoint, ref acquiredConnectionLock);
-
-                    if (entry.Connections.Length >= this.connectionOptions.ConnectionsPerEndpoint)
+                    if (entry.PendingConnection != null
+                        && entry.Connections.Length < this.connectionOptions.ConnectionsPerEndpoint)
                     {
-                        result = entry.Connections;
-                        break;
+                        // Start a new connection attempt in the background.
+                        this.GetConnectionAsync(endpoint).Ignore();
                     }
```

Comment:
Previously, when a `GetConnection(SiloAddress)` call was received, it would always try to create a new connection if we had less than `ConnectionsPerEndpoint` connections. That's not optimal since it stalls more messages than are needed. 

Instead we will now fire off an async attempt to get a connection and return an established connection synchronously.